fe/fe-core/src/main/java/com/starrocks/sql/optimizer/CTEUtils.java

```diff
@@ -25,70 +24,50 @@
+    public static void collectCteOperatorsImpl(OptExpression root, OptimizerContext context, BitSet searchConsume) {
+        if (OperatorType.LOGICAL_CTE_ANCHOR.equals(root.getOp().getOpType())) {
             // produce
-            LogicalCTEProduceOperator produce = (LogicalCTEProduceOperator) root.getOp();
-            context.getCteContext().addCTEProduce(produce.getCteId());
-        }
-    }
+            LogicalCTEAnchorOperator anchor = (LogicalCTEAnchorOperator) root.getOp();
+            context.getCteContext().addCTEProduce(anchor.getCteId());
 
-    private static void collectCteConsume(OptExpression root, OptimizerContext context) {
-        for (Integer cteId : context.getCteContext().getAllCTEProduce()) {
-            OptExpression anchor = findCteAnchor(root, cteId);
-            Preconditions.checkNotNull(anchor);
-            collectCteConsumeImpl(anchor, cteId, context);
-        }
-    }
+            Preconditions.checkState(root.getInputs().get(0).getOp() instanceof LogicalCTEProduceOperator);
+            LogicalCTEProduceOperator produce = (LogicalCTEProduceOperator) root.getInputs().get(0).getOp();
+            Preconditions.checkState(produce.getCteId() == anchor.getCteId());
 
-    private static OptExpression findCteAnchor(OptExpression root, Integer cteId) {
-        LinkedList<OptExpression> queue = Lists.newLinkedList();
-        queue.addLast(root);
+            BitSet left = (BitSet) searchConsume.clone();
+            collectCteOperatorsImpl(root.inputAt(0), context, left);
 
-        while (!queue.isEmpty()) {
-            OptExpression expression = queue.pollFirst();
-            if (OperatorType.LOGICAL_CTE_ANCHOR.equals(expression.getOp().getOpType()) &&
-                    ((LogicalCTEAnchorOperator) expression.getOp()).getCteId() == cteId) {
-                return expression.getInputs().get(1);
-            }
-
-            expression.getInputs().forEach(queue::addLast);
+            BitSet right = (BitSet) searchConsume.clone();
+            right.set(anchor.getCteId());
+            collectCteOperatorsImpl(root.inputAt(1), context, right);
+            return;
         }
-        return null;
-    }
 
-    private static void collectCteConsumeImpl(OptExpression root, Integer cteId, OptimizerContext context) {
         if (OperatorType.LOGICAL_CTE_CONSUME.equals(root.getOp().getOpType())) {
-            if (((LogicalCTEConsumeOperator) root.getOp()).getCteId() != cteId) {
-                // not ask children
-                return;
-            }
             // consumer
             LogicalCTEConsumeOperator consume = (LogicalCTEConsumeOperator) root.getOp();
-            context.getCteContext().addCTEConsume(consume.getCteId());
-            // not ask children
+
+            if (searchConsume.get(consume.getCteId())) {
+                context.getCteContext().addCTEConsume(consume.getCteId());
+                searchConsume = new BitSet();
```

Comment:
Reset here? Does this means cteAnchor is illegal only if consume is found?

Reply:
no, in a complex SQL 
```
with x as (
    select * from t1
),
y as (
select * from y 
)
select * from y;
```
the plan:
```
         Anchor(x)          
       /          \  
  produce(x)      Anchor(y)        
 /               /        \
x         produce(y)    consume(y)
           /                \
     consume1(x)           consume2(x)
        /                     \
       x                       x
```

I only count consume1(X) which one under produce(y), don't count consume2(X) under consume(y), because the existence of consume2 depends on consume(Y). if CTE(Y) inline, count consume2 is meaningless